### ISSUE TITLE
Problem with depth-image

### ISSUE THREAD

**Issue by KatharinaSchmidt** [j96w/DenseFusion#174]
Hello, 
I try to train DenseFusion on my own dataset. But I seem to have a problem with the depth-image and the pointcloud which is generated from it by 
``` ruby
cam_scale = 1.0
pt2 = depth_masked / cam_scale
pt0 = (ymap_masked - self.cam_cx) * pt2 / self.cam_fx
pt1 = (xmap_masked - self.cam_cy) * pt2 / self.cam_fy
cloud = np.concatenate((pt0, pt1, pt2), axis=1)
cloud = cloud / 1000.0
```
My depth image is in mm 16 bit (synthesized with NDDS). 
A projection of the point-cloud into the color-image: 
![cloud](https://user-images.githubusercontent.com/62353042/91174629-d42c7400-e6df-11ea-9ea4-1de0cb67d58c.png)
As you can see the objectshape seems to be correct, but the pose is wrong.
I checked the corresponding depth image, there the pose is correct. 

I also have a question about the values in the array from depth-image. In the code we take the depth-image as an array with  `depth = np.array(Image.open(self.list_depth[index]))` . If I print the values of my own depth-image it starts with 1100, which should be the depth in mm. But if I print the same values for a depth-image from linemod the array starts with 0, which would mean a depth of 0 mm.
Can you tell me why?

**Comment by Ixion46** [comment]
I think the linemod data was recorded with a kinect and in my opinion, the kinect give depth images in meters. So I think thats why the array there has much smaller values.

At the readme of densefusion is also written :  Please make sure that the distance metric in your dataset should be converted to meter, otherwise the hyperparameter w need to be adjusted.

**Comment by Ixion46** [comment]
Hey @KatharinaSchmidt were you able to train your network? My Training Session does not start and I dont know what the problem is, as I mentioned in my new issue.

**Comment by Ixion46** [comment]
Hey @KatharinaSchmidt  , do you already have trained your network? What is your avg_dis that it achieves? Mine is actual at around 7cm, what isnt the best unfortunately.

I am thinking that my transformations from the json files are wrong. How did you calculated the poses matrices from the json files?

**Comment by KatharinaSchmidt** [comment]
My average distance is around 3cm I think. For the pose I used the location values given in json-file in cm and transfered them to millimeter. For the rotation you have to be aware of the fact, that NDDS annotates pose in lefthandside coordinatesystem, but DenseFusion uses righthandside coordinatesystem.

**Comment by Ixion46** [comment]
Yeah I did all that, but i still dont get close to 2cm. Which length had your objects when you created the pictures with the unreal engine? You had the objects in m , cm or mm ? Thanks for all your help

**Comment by Ixion46** [comment]
I always get errors while trying to project the points of my .mat files into the color image. Can you @KatharinaSchmidt  help me a last time and post the codefragment that you used for it?
Thankfully Ixion

**Comment by KatharinaSchmidt** [comment]
I didn't use a specific .mat file. 
In my `dataset.py` I just read the informations out of the .json file:
``` ruby
# open the annotation-file
with open(self.list_json[index]) as jsonfile:
            object_data = json.load(jsonfile)

# get the position from annotations
target_t = np.array(object_data['objects'][0]['location']) * 10 # NDDS gives units in centimeters

# get the rotation from annotation file
rot= np.array(object_data['objects'][0]['pose_transform'])[0:3,0:3]

# get the corners of bounding boxes from annotations
bbox = object_data['objects'][0]['bounding_box']
rmax, cmax = np.asarray(bbox['top_left'])
rmin, cmin = np.asarray(bbox['bottom_right'])
rmax, rmin, cmax, cmin = int(rmin), int(rmax), int(cmin), int(cmax)
```
with this code you get the pose in correct units to train densefusion with linemod-code.

**Comment by Ixion46** [comment]
With that Code, I get the same rotation error as you posted here https://github.com/j96w/DenseFusion/issues/76 .

**Comment by akeaveny** [comment]
Hi @KatharinaSchmidt, 

I'm looking into DenseFusion again and noticed the same ab my depth images as to what you have above.
I've looked into YCB, Linemod and my NDDS depth images. Here are the min and maxs.

```
YCB       Min: 0             Max: 48210
Linemod   Min: 0             Max: 6497
NDDS      Min:  436          Max: 65535 [???]
```

I find it weird that NDDS depth images don't start at 0. I used the 16 bit mm depth images (i.e. .depth.mm.16.png).
The max value of is ~65m which doesn't seem right.

I wonder if you did any preprocessing to your depth images like [this](https://github.com/NVIDIA/Dataset_Synthesizer/issues/10) (i.e. MAX_DEPTH / (2 ** 16)).
I've made sure my units match so I can use the Linemod format but can't get my loss below 0.0575 [m] or 5.75 cm.
I changed my opt.w param to 0.05 to have my w's close to 1.

```
        cam_scale = 1e3 # now in [m]
        pt2 = depth_masked / cam_scale
        pt0 = (ymap_masked - cam_cx) * pt2 / cam_fx
        pt1 = (xmap_masked - cam_cy) * pt2 / cam_fy
        cloud = np.concatenate((pt0, pt1, pt2), axis=1)
        if self.add_noise:
            cloud = np.add(cloud, translation_noise)
```

**Comment by Ixion46** [comment]
Hey @akeaveny I also observed that behaviour of NDDS, so I converted the depth-values above a certain threshold to zero. Just use a threshold value where you are sure, that no object is there. I have a avg_dis of 1,5cm at the moment. Some objects work really good with a distance of 0.3 cm and some others dont work at all dis > 3cm.
I hope I was able to help you

**Comment by an99990** [comment]
is it normal that pt0 uses ymap and not xmap ? why are they reversed ?

```
pt0 = (ymap_masked - cam_cx) * pt2 / cam_fx
pt1 = (xmap_masked - cam_cy) * pt2 / cam_fy
```

**Comment by an99990** [comment]
if depth map is in meters, why are they dividing by 1000 ? @huckl3b3rry87 @akeaveny @KatharinaSchmidt . My mesh are in meters so i removed the /1000, should i bring it back ?

```py
        cloud = cloud / 1000.0

        model_points = self.pt[obj] / 1000.0
        dellist = [j for j in range(0, len(model_points))]
        dellist = random.sample(dellist, len(model_points) - self.num_pt_mesh_small)
        model_points = np.delete(model_points, dellist, axis=0)

        target_r = np.resize(np.array(meta['cam_R_m2c']), (3, 3))
        target_t = np.array(meta['cam_t_m2c'])
        add_t = np.array([random.uniform(-self.noise_trans, self.noise_trans) for i in range(3)])

        if self.add_noise:
            cloud = np.add(cloud, add_t)

        target = np.dot(model_points, target_r.T)
        if self.add_noise:
            target = np.add(target, target_t / 1000.0 + add_t)
            out_t = target_t / 1000.0 + add_t
        else:
            target = np.add(target, target_t / 1000.0)
            out_t = target_t / 1000.0
```
